src/software/ai/passing/cost_function.cpp

```diff
@@ -8,70 +8,66 @@
-        ratePassShootScore(world.field(), world.enemyTeam(), pass, passing_config);
+    double shoot_pass_rating = ratePassShootScore(
+        world.ball(), world.field(), world.enemyTeam(), pass, passing_config);
 
-    // Rate all passes outside our target region as 0 if we have one
-    double in_region_quality = 1;
-    if (target_region)
-    {
-        in_region_quality = rectangleSigmoid(*target_region, pass.receiverPoint(), 0.1);
-    }
-
-    // Place strict limits on pass start time
-    double min_pass_time_offset =
-        passing_config->getMinTimeOffsetForPassSeconds()->value();
-    double max_pass_time_offset =
-        passing_config->getMaxTimeOffsetForPassSeconds()->value();
-    double pass_time_offset_quality =
-        sigmoid(pass.startTime().toSeconds(),
-                min_pass_time_offset + world.getMostRecentTimestamp().toSeconds(), 0.5) *
-        (1 - sigmoid(pass.startTime().toSeconds(),
-                     max_pass_time_offset + world.ball().timestamp().toSeconds(), 0.5));
+    double in_region_quality = rectangleSigmoid(zone, pass.receiverPoint(), 0.2);
 
     // Place strict limits on the ball speed
     double min_pass_speed     = passing_config->getMinPassSpeedMPerS()->value();
     double max_pass_speed     = passing_config->getMaxPassSpeedMPerS()->value();
     double pass_speed_quality = sigmoid(pass.speed(), min_pass_speed, 0.2) *
                                 (1 - sigmoid(pass.speed(), max_pass_speed, 0.2));
 
-    // We ignore shoot pass quality if we're rating for a "RECEIVE_AND_DRIBBLE" pass
-    double pass_quality = 0;
-    switch (pass_type)
+    return static_pass_quality * friendly_pass_rating * enemy_pass_rating *
+           shoot_pass_rating * pass_speed_quality * in_region_quality;
+}
+
+double rateZone(const Field& field, const Rectangle& zone, const Point& ball_position,
+                std::shared_ptr<const PassingConfig> passing_config)
+{
+    // If the zone contains the ball, its not a good place to receive a pass
+    if (contains(zone, ball_position))
     {
-        case PassType::RECEIVE_AND_DRIBBLE:
-            pass_quality = static_pass_quality * friendly_pass_rating *
-                           enemy_pass_rating * in_region_quality *
-                           pass_time_offset_quality * pass_speed_quality;
-            break;
-        case PassType::ONE_TOUCH_SHOT:
-            pass_quality = static_pass_quality * friendly_pass_rating *
-                           enemy_pass_rating * shoot_pass_rating * in_region_quality *
-                           pass_time_offset_quality * pass_speed_quality;
-            break;
-        default:
-            throw std::invalid_argument("Unhandled pass type given to `ratePass`");
+        return 0.0;
     }
 
-    return pass_quality;
+    // Zones with their centers in bad positions are not good
+    double static_pass_quality =
+        getStaticPositionQuality(field, zone.centre(), passing_config);
```

Comment:
Did you consider doing some sort of function over several points in the zone (max/min/mean/whatever) for all the cost functions here?

Reply:
Just adding another idea here, how I thought this function might originally be implemented would be:

If you are keeping at least 1 point in each zone as you optimize, just take the rating/score of this point to determine the score of the zone. Because it's being optimized you can assume it will generally be tracking the best point in the zone, and therefore be the best representation of the zone's max rating at a given time.

Reply:
Def did consider probing multiple points of the zone to see if they rank well, the zones are small enough that seems to work well for now. If this is impossible to generalize to all pitch divisions, will consider pulling it out of cost_functions.cpp

The friendly capability score in the zones without robots is 0 so we can't really use the best  pass in zone as a cherry pick location (it would just be a random score very close to 0 if not 0). 

If you take out the friendly capability score, the rest of the scores combined isn't enough information to encourage passes that are long enough to make progress up the field. 

This kind of also relates to the work needed to be done here:  https://github.com/UBC-Thunderbots/Software/pull/1980#discussion_r602972849